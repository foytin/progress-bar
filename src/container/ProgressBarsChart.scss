.progress-options {
  margin-bottom: 1rem;
}

.loading {
  margin-top: 2rem;

  .loading-icon {
    margin-right: 1rem;
  }

  .loading-text {
    font-size: 2rem;
    color: #555;
  }
}