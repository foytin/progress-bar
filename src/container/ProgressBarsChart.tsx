import React, { useState, useEffect } from "react";
import axios from "axios";

import ProgressBar from "../component/ProgressBar";
import StepButton from "../component/StepButton";
import BarSelection from "../component/BarSelection";
import Error from "../component/Error";

import { isVaildArray, keepUniqueArrayValue, sortArray } from "../util/index";

import styles from "./ProgressBarsChart.scss";

const ProgressBarsChart = (): React.ReactElement => {
  const [bars, setBars] = useState<number[]>([]);
  const [buttons, setButtons] = useState<number[]>([]);
  const [threshold, setThreshold] = useState<number>(0);
  const [barIndex, setBarIndex] = useState<string>("0");
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<boolean>(false);

  useEffect(() => {
    axios
      .get("http://pb-api.herokuapp.com/bars")
      .then((res) => {
        const { data } = res;
        const barsArr = data.bars || [];
        let buttonArr = data.buttons || [];
        const limit = data.limit || 100;
        if (isVaildArray(barsArr)) {
          setBars(barsArr);
        } else {
          setBars([]);
        }
        if (isVaildArray(buttonArr)) {
          // remove duplicated step value
          buttonArr = keepUniqueArrayValue(buttonArr);
          // sort buttons based on button value
          sortArray(buttonArr);
          setButtons(keepUniqueArrayValue(buttonArr));
        } else {
          setButtons([]);
        }
        setThreshold(limit);
        setLoading(false);
      })
      .catch((err) => {
        console.log(err);
        setError(true);
        setLoading(false);
      });
  }, []);

  const changeProgress = (step: number): void => {
    let value: number | null = null;
    const updatedBars: Array<number> = [...bars];

    if (updatedBars[barIndex] + step <= 0) {
      value = 0;
    } else {
      value = updatedBars[barIndex] + step;
    }
    updatedBars[barIndex] = value;
    setBars([...updatedBars]);
  };

  const loadingPage = (): JSX.Element => (
    <div className={styles.loading} data-testid="loading-spin">
      <span
        className={`spinner-border text-primary ${styles["loading-icon"]}`}
        role="status"
      />
      <span className={styles["loading-text"]}>Data loading...</span>
    </div>
  );

  return (
    <div>
      <h1>Progress Bars Demo</h1>
      {loading ? (
        loadingPage()
      ) : (
        <>
          {bars.map((val, index) => (
            <ProgressBar key={`bar${index}`} value={val} limit={threshold} />
          ))}
          <section className="row">
            <div className={`col-md-5 col-sm-12 ${styles["progress-options"]}`}>
              <BarSelection bars={bars} onChange={(val) => setBarIndex(val)} />
            </div>
            <div className="col-md-7 col-sm-12">
              {buttons.map((val, index) => (
                <StepButton
                  key={`button${index}`}
                  step={val}
                  onClick={changeProgress}
                />
              ))}
            </div>
          </section>
        </>
      )}
      {error ? <Error /> : null}
    </div>
  );
};

export default ProgressBarsChart;
