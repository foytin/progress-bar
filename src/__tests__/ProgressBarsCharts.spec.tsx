import React from "react";
import { act } from "react-dom/test-utils";
import axios from "axios";
import {
  render,
  fireEvent,
  waitFor,
  RenderResult,
} from "@testing-library/react";

import ProgressBarsChart from "../container/ProgressBarsChart";

describe("ProgressBarsChart container intergation test cases", () => {
  const data = { buttons: [7, 33, -34], bars: [60, 43, 24, 33], limit: 90 };

  it("should has lodaing spin in inital status", () => {
    const wrapper = render(<ProgressBarsChart />);
    expect(wrapper.getByTestId("loading-spin")).toBeInTheDocument();
    expect(wrapper.findByText("Data loading...")).toBeTruthy();
  });

  it("should have 4 bars, 3 buttons and 4 bar options after grab api data", async () => {
    let wrapper: RenderResult;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });

    await waitFor(() => wrapper.getAllByTestId("progress-bar"));
    expect(wrapper!.container).toMatchSnapshot();
    expect(wrapper!.getAllByTestId("progress-bar")).toHaveLength(4);
    expect(wrapper!.getAllByRole("option")).toHaveLength(4);
    expect(wrapper!.getAllByTestId("step-btn")).toHaveLength(3);
  });

  it("After grab api data, first bar should display 60% with class is normal and style is width: 60%", async () => {
    let wrapper: RenderResult;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });

    await waitFor(() => wrapper.getAllByTestId("progress-bar"));
    expect(wrapper!.getAllByTestId("bar-number")[0]).toHaveTextContent(
      `${data.bars[0]}%`
    );
    expect(wrapper!.getAllByTestId("bar-block")[0]).toHaveClass("bar normal");
    expect(wrapper!.getAllByTestId("bar-block")[0]).toHaveStyle(
      `width: ${data.bars[0]}%`
    );
  });

  it("select second bar option and click second step btn, second bar value will increase 7 from 43 to 50", async () => {
    let wrapper: RenderResult;
    const buttons = [...data.buttons].sort((a, b) => a - b);
    const selectionIndex = 1;
    const btnIndex = 1;
    const barIndex = 1;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });

    await waitFor(() => wrapper.getAllByTestId("progress-bar"));

    act(() => {
      fireEvent.blur(wrapper.getByTestId("bar-selection"), {
        target: { value: selectionIndex },
      });
    });

    act(() => {
      fireEvent.click(wrapper.getAllByTestId("step-btn")[btnIndex]);
    });

    const barOption = wrapper!.container
      .getElementsByTagName("option")
      .item(selectionIndex);
    expect(barOption!.selected).toBeTruthy();

    expect(wrapper!.getAllByTestId("bar-number")[barIndex]).toHaveTextContent(
      `${data.bars[barIndex] + buttons[btnIndex]}%`
    );
  });

  it("value below threshold value, porgress bar should have normal classname, while equal or above should have alert classname", async () => {
    let wrapper: RenderResult;
    const barIndex = 0;
    const btnIndex = 2;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });

    await waitFor(() => wrapper.getAllByTestId("progress-bar"));

    expect(wrapper!.getAllByTestId("bar-number")[barIndex]).toHaveTextContent(
      "60%"
    );
    expect(wrapper!.getAllByTestId("bar-block")[barIndex]).toHaveClass(
      "bar normal"
    );

    act(() => {
      fireEvent.click(wrapper.getAllByTestId("step-btn")[btnIndex]);
    });

    // threshold value is 90%
    expect(wrapper!.getAllByTestId("bar-number")[barIndex]).toHaveTextContent(
      "93%"
    );
    expect(wrapper!.getAllByTestId("bar-block")[barIndex]).toHaveClass(
      "bar alert"
    );
  });

  it("Progress bar minimal value should be equal or above 0, even calculate result is negative value", async () => {
    let wrapper: RenderResult;
    const barIndex = 2;
    const selectionIndex = 2;
    const btnIndex = 0;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });

    await waitFor(() => wrapper.getAllByTestId("progress-bar"));

    act(() => {
      fireEvent.blur(wrapper.getByTestId("bar-selection"), {
        target: { value: selectionIndex },
      });
    });

    act(() => {
      fireEvent.click(wrapper.getAllByTestId("step-btn")[btnIndex]);
    });
    // calculate result is -10
    expect(wrapper!.getAllByTestId("bar-number")[barIndex]).toHaveTextContent(
      "0%"
    );
  });
});

describe("ProgressBarsChart container intergation exception test cases", () => {
  it("should return empty array if data error", async () => {
    const data = { buttons: undefined, bars: undefined, limit: undefined };
    let wrapper: RenderResult;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });
    await waitFor(() => wrapper.getAllByTestId("bar-selection"));
    expect(wrapper!.queryAllByTestId("progress-bar")).toEqual([]);
    expect(wrapper!.queryAllByRole("option")).toEqual([]);
    expect(wrapper!.queryAllByTestId("step-btn")).toEqual([]);
  });

  it("should return empty array if data.bars and data.buttons is not array", async () => {
    const data = { buttons: {}, bars: {}, limit: 100 };
    let wrapper: RenderResult;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest
        .spyOn(axios, "get")
        .mockImplementation(() => Promise.resolve({ data }));
    });
    await waitFor(() => wrapper.getAllByTestId("bar-selection"));
    expect(wrapper!.queryAllByTestId("progress-bar")).toEqual([]);
    expect(wrapper!.queryAllByRole("option")).toEqual([]);
    expect(wrapper!.queryAllByTestId("step-btn")).toEqual([]);
  });

  it("should display Error page when call api fail", async () => {
    let wrapper: RenderResult;
    act(() => {
      wrapper = render(<ProgressBarsChart />);
      jest.spyOn(axios, "get").mockRejectedValueOnce("network failure");
    });

    await waitFor(() => wrapper.getByTestId("error-page"));
    expect(wrapper!.queryByTestId("loading-spin")).toEqual(null);
    expect(wrapper!.getByTestId("error-page")).toBeInTheDocument();
  });
});
