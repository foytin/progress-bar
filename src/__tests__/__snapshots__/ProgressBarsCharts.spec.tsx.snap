// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarsChart container intergation test cases should have 4 bars, 3 buttons and 4 bar options after grab api data 1`] = `
<div>
  <div>
    <h1>
      Progress Bars Demo
    </h1>
    <div
      class="container"
      data-testid="progress-bar"
    >
      <p
        data-testid="bar-number"
      >
        60%
      </p>
      <div
        class="bar normal"
        data-testid="bar-block"
        style="width: 60%;"
      />
    </div>
    <div
      class="container"
      data-testid="progress-bar"
    >
      <p
        data-testid="bar-number"
      >
        43%
      </p>
      <div
        class="bar normal"
        data-testid="bar-block"
        style="width: 43%;"
      />
    </div>
    <div
      class="container"
      data-testid="progress-bar"
    >
      <p
        data-testid="bar-number"
      >
        24%
      </p>
      <div
        class="bar normal"
        data-testid="bar-block"
        style="width: 24%;"
      />
    </div>
    <div
      class="container"
      data-testid="progress-bar"
    >
      <p
        data-testid="bar-number"
      >
        33%
      </p>
      <div
        class="bar normal"
        data-testid="bar-block"
        style="width: 33%;"
      />
    </div>
    <section
      class="row"
    >
      <div
        class="col-md-5 col-sm-12 progress-options"
      >
        <select
          class="custom-select custom-select-lg"
          data-testid="bar-selection"
        >
          <option
            value="0"
          >
            #Progress 1
          </option>
          <option
            value="1"
          >
            #Progress 2
          </option>
          <option
            value="2"
          >
            #Progress 3
          </option>
          <option
            value="3"
          >
            #Progress 4
          </option>
        </select>
      </div>
      <div
        class="col-md-7 col-sm-12"
      >
        <button
          class="btn btn btn-lg btn-outline-dark"
          data-testid="step-btn"
          type="button"
        >
          -34
        </button>
        <button
          class="btn btn btn-lg btn-outline-dark"
          data-testid="step-btn"
          type="button"
        >
          +7
        </button>
        <button
          class="btn btn btn-lg btn-outline-dark"
          data-testid="step-btn"
          type="button"
        >
          +33
        </button>
      </div>
    </section>
  </div>
</div>
`;
